File: r/computerAB.R
#!/usr/bin/Rscript
getA <- function(str1, str2) {
    len1 <- nchar(str1)
    len2 <- nchar(str2)
    result <- 0

    if (len1 != len2) {
        return(0)
    }

    result <- 0
    for (i in c(1:4)) {
        if (substring(str1, i, i) == substring(str2, i, i)) {
            result <- result + 1
        }
    }

    return(result)
}

getB <- function(str1, str2) {
    len1 <- nchar(str1)
    len2 <- nchar(str2)
    result <- 0

    if (len1 != len2) {
        return(0)
    }

    result <- 0
    for (i in c(1:4)) {
        for (j in c(1:4)) {
            if (i != j) {
                if (substring(str1, i, i) == substring(str2, j, j)) {
                    result <- result + 1
                }
            }
        }
    }

    return(result)
}

results <- c()
for (i in c(0:9)) {
    for (j in c(0:9)) {
        for (k in c(0:9)) {
            for (m in c(0:9)) {
                if (i != j && i != k && i != m && j != k && j != m && k != m) {
                    cresult <- paste(toString(i), toString(j), toString(k), toString(m), sep = "")
                    results <- c(results, cresult)
                }
            }
        }
    }
}
item <- sample(length(results), 1)
answer <- results[item]
print(paste("My answer: ", answer))

repeat {
    cat("The A value is: ");
    con <- file("stdin")
    avalue <- readLines(con, n = 1)
    close(con)

    cat("The B value is: ");
    con <- file("stdin")
    bvalue <- readLines(con, n = 1)
    close(con)

    if (avalue == 4 && bvalue == 0) {
        print("Game over.")
        break
    }

    newresults <- c()
    for (value in results) {
        A <- getA(answer, value)
        B <- getB(answer, value)
        if (A == avalue && B == bvalue) {
            newresults <- c(newresults, value)
        }
    }

    results <- newresults

    if (length(results) > 0) {
        answer <- results[1]
        print(paste("My answer: ", answer))
    } else {
        print("Something is wrong!!!")
        break
    }
}

File: php/computerAB.php
#!/usr/bin/env php
<?php

function getA(string $answer, string $guess): int {
    $num = 0;
    $len = strlen($answer);
    if($len != strlen($guess)) {
        return 0;
    }

    for ($count = 0; $count < $len; $count++) {
        if ($answer[$count] == $guess[$count]) {
            $num++;
        }
    }

    return $num;
}

function getB(string $answer, string $guess): int {
    $num = 0;

    $len1 = strlen($answer);
    $len2 = strlen($guess);
    if($len1 != $len2) {
        return 0;
    }

    for ($count = 0; $count < $len1; $count++) {
        for ($count2 = 0; $count2 < $len2; $count2++) {
            if ($answer[$count] == $guess[$count2]) {
                if ($count != $count2) {
                    $num++;
                }
            }
        }
    }

    return $num;
}


$resultset= [];

for ($i = 0; $i < 10; $i++) {
    for ($j = 0; $j < 10; $j++) {
        for ($k = 0; $k < 10; $k++) {
            for ($m = 0; $m < 10; $m++) {
                if ($i != $j && $i != $k && $i != $m && $j != $k && $j != $m && $m != $k) {
                    $result = (string )$i . (string)$j . (string )$k . (string )$m;
                    array_push($resultset, $result);
                }
            }
        }
    }
}

while (1) {
     if (sizeof($resultset) == 0) {
         exit("Game over, you give the wrong answer.\n");
     }

     $rand_number = rand(0, sizeof($resultset) - 1);
     $current = $resultset[$rand_number];
     echo ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n";
     echo "My guess number is " . $current . ".\n";
     $a = (int) readline("The A is: ");
     $b = (int) readline("The B is: ");

     if($a < 0 || $a > 4 || !is_integer($a)) {
          exit("Wrong answer!!!\n");
     }

     if($b < 0 || $b > 4 || !is_integer($b)) {
          exit("Wrong answer!!!\n");
     }

     if($a == 4 && $b == 0) {
          echo "The game is completed.\n";
          break;
     }

     foreach ($resultset as $r) {
          $guessa = getA($current, $r);
          $guessb = getB($current, $r);

          if ($guessa != $a || $guessb != $b) {
              $key = array_search($r, $resultset);
              if ($key !== false) {
                   unset($resultset[$key]);
              }
          }
     }

     $resultset = array_values($resultset);
}
?>
